scripts/signing_kernel_modules.sh

```diff
@@ -0,0 +1,124 @@
+
+Parameters description:
+LINUX_KERNEL_VERSION
+PEM_CERT                             public key (pem format)
+PEM_PRIVATE_KEY                      private key (pem format)
+LOCAL_SIGN_FILE                      path of the sign-file tool for signing Kernel Modules, if the value is empty it will used the sign-file installed in /usr/lib/linux-kbuild-<version>/scripts
+LOCAL_EXTRACT_CERT                   path of the extract-cert tool for Extract X.509 certificate, if the value is empty it will used the extract-cert installed in /usr/lib/linux-kbuild-<version>/scripts
+KERNEL_MODULES_DIR                   root directory of all the kernel modules to be sign by the script, if the value empty it will use the call script location as root.
+
+Runs examples:
+1. sudo bash scripts/signing_kernel_modules.sh 5.10.0-8-2 cert.pem priv-key.pem
+2. sudo bash scripts/signing_kernel_modules.sh 5.10.0-8-2 cert.pem priv-key.pem fsroot-mellanox /usr/lib/linux-kbuild-5.10/scripts/extract-cert /usr/lib/linux-kbuild-5.10/scripts/sign-file
+EOF
+}
+
+while getopts 'l:c:p:k:s:e:hv' flag; do
+  case "${flag}" in
+    l) LINUX_KERNEL_VERSION="${OPTARG}" ;;
+    c) PEM_CERT="${OPTARG}" ;;
+    p) PEM_PRIVATE_KEY="${OPTARG}" ;;
+    k) KERNEL_MODULES_DIR="${OPTARG}" ;;
+    s) LOCAL_SIGN_FILE="${OPTARG}" ;;
+    e) LOCAL_EXTRACT_CERT="${OPTARG}" ;;
+    v) VERBOSE='true' ;;
+    h) print_usage
+       exit 1 ;;
+  esac
+done
+if [ $OPTIND -eq 1 ]; then echo "no options were pass"; print_usage; exit 1 ;fi
+
+if [ -z ${LINUX_KERNEL_VERSION} ]; then
+    echo "ERROR: LINUX_KERNEL_VERSION arg1 is empty"
+    usage
+    exit 1
+fi
+
+if [ ! -f ${PEM_CERT} ]; then
+    echo "ERROR: arg2 PEM_CERT=${PEM_CERT} file does not exist"
+    usage
+    exit 1
+fi
+
+if [ ! -f ${PEM_PRIVATE_KEY} ]; then
+    echo "ERROR: arg3 PEM_PRIVATE_KEY=${PEM_PRIVATE_KEY} file does not exist"
+    usage
+    exit 1
+fi
+
+kbuild_ver_major="$(cut -d '.' -f 1 <<< "$LINUX_KERNEL_VERSION")"."$(cut -d '.' -f 2 <<< "$LINUX_KERNEL_VERSION")"
+if [ -z ${LOCAL_SIGN_FILE} ]; then
+    LOCAL_SIGN_FILE="/usr/lib/linux-kbuild-${kbuild_ver_major}/scripts/sign-file"
+fi
+
+if [ ! -f ${LOCAL_SIGN_FILE} ]; then
+    echo "ERROR: LOCAL_SIGN_FILE=${LOCAL_SIGN_FILE} file does not exist"
+    usage
+    exit 1
+fi
+
+if [ -z ${LOCAL_EXTRACT_CERT} ]; then
+    LOCAL_EXTRACT_CERT="/usr/lib/linux-kbuild-${kbuild_ver_major}/scripts/extract-cert"
+fi
+
+if [ ! -f ${LOCAL_EXTRACT_CERT} ]; then
+    echo "ERROR: LOCAL_EXTRACT_CERT=${LOCAL_EXTRACT_CERT} file does not exist"
+    usage
+    exit 1
+fi
+
+if [ ! -d "$KERNEL_MODULES_DIR" ]; then
+  # If the user do not provide a KERNEL_MODULES_DIR, the script is going to search in the script call path for Kernel modules.
+  KERNEL_MODULES_DIR="./"
+  echo "KERNEL_MODULES_DIR set to default path: $KERNEL_MODULES_DIR"
+
+fi
+
+# find all the kernel modules.
+modules_list=$(sudo find ${KERNEL_MODULES_DIR} -name "*.ko")
```

Comment:
what is the kernel_module_dir scattered in multiple places? i also remember bash has some limitation on the length of maximum arguments, i doubt this will work.

Reply:
this code is working for sure, it was already tested on different devices.
The reason that the code is looking for all the kernel object by using "find", is to make the signing flow more robust, that because each vendor have its own kernel modules and it is not feasible to save all the kernel module's path. So by finding all the kernel modules, if any vendor add some new kernel module the code will find it and sign it as well.

Reply:
in the build image, it should have multiple kernel module dir, can KERNEL_MODULES_DIR be a list?

Reply:
the KERNEL_MODULES_DIR is assigned to be the file system root, and we are searching all the kernel objects inside this directory.

Because the search is recursive, it's like searching in multiples list dir.
It's literally searching in all the file system.